### PR TITLE
Some broadcast fixes

### PR DESCRIPTION
**Pull request by p-zubieta**
Fixes for #17984 and #18176

Cc: @stevengj

### CODE REVIEW

File: test/broadcast.jl [hunk 1]
```diff
@@ -304,9 +304,21 @@ end
...
+    @test eltype(broadcast(f17314, Int[])) === Any
 end
 let io = IOBuffer()
     broadcast(x->print(io,x), 1:5) # broadcast with side effects
     @test takebuf_array(io) == [0x31,0x32,0x33,0x34,0x35]
 end
+
+# Issue 18176
+let f18176(a, b, c) = a + b + c
+    @test f18176.(1.0:2, 3, 4) == f18176.(3.0, 1.0:2, 4.0) == broadcast(f18176, 3, 4, 1.0:2)
```
Comment (stevengj) [inline test/broadcast.jl]: Not that it matters because `+` is commutative, but for consistency with the dot calls this should be `broadcast(f18176, 3, 1.0:2,  4)`

File: base/number.jl [hunk 2]
```diff
@@ -63,4 +63,6 @@ zero{T<:Number}(::Type{T}) = convert(T,0)
 one(x::Number)  = oftype(x,1)
 one{T<:Number}(::Type{T}) = convert(T,1)
 
+_default_type(::Type{Number}) = Int
```
Comment (stevengj) [inline base/number.jl]: What is this change for?  Note that `_default_type(T::Union{Type{Real},Type{AbstractFloat}}) = Float64` in `float.jl` ... it seems a bit odd that `_default_type(Real) == Float64` but `_default_type(Number) == Int`.

Comment (pabloferz) [inline base/number.jl]: This was missing for preserving containers of `Number` as `eltype`. When the container is of an abstract type, `promote_op` uses `_default_type` and then the type found is '`typejoin`ed with the `eltype` of the container. I guess I could use `Float64`, but I thought that `Int` should give a reasonable result (that is, not `Any`) generally.

Comment (stevengj) [inline base/number.jl]: What's wrong with leaving it as the default (`Number`) rather than returning a concrete type?

Comment (pabloferz) [inline base/number.jl]: If you have `Base._promote_op(+, Int, Number)` that will return `Any`, so what `promote_op` currently does is to call `Base._promote_op(+, Int, _default_type(Number))` which with this would return `Int` and then it returns `typejoin(Int, Number, Int) == Number` (the types of the inputs and the sample return type).

For `Number` it seems that we could let the container as `Number`, but for other containers that might not be the best option. If you do `broadcast(*, 1.5, Integer[1])`, `Integer` as the final container wouldn't be right , so `promote_op` does `Base._promote_op(*, Float64, Int)` (that is `Float64`) and then returns `Real` (by `typejoin`ing these).

Comment (pabloferz) [inline base/number.jl]: But I can change it to `Float64`, in the end I don't think it would matter too much for the case of `Number`.

Comment (stevengj) [inline base/number.jl]: I would prefer to leave it just as `Number`.  Having `Base.promote_op(+, Int, Number) == Number`  seems fine to me.

Moreover, this seems unrelated to #17984 and #18176 ... in general, @pabloferz, it makes it much harder to review PRs if you include unrelated changes in the PR.

Comment (pabloferz) [inline base/number.jl]: Sorry. I'll leave this change out of the PR. Or should I just put it un another commit?

Comment (stevengj) [inline base/number.jl]: Out of the PR

Comment (stevengj) [inline base/number.jl]: (I would make the same recommendation for your other PRs. e.g. separate the PR to add tuple support to broadcast from the PR to generalize the concept of scalar.)

**Comment by stevengj** [comment]
Restarted CI, since the Travis failure seems like it could have been just a glitch.

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18200#issuecomment-243129233) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6f467b026b4ca6db0d105a33e8ce9d75a338dd78/b9bfa2c_vs_6d179b3/report.md). cc @jrevels
